### PR TITLE
(WIP) Add avr plrule to manage user plrule

### PR DESCRIPTION
**Pull request by BenamarMk**
Controlling subscription placement decisions through AVR own placement rule.
For details see: https://hackmd.io/4Lzuhb4lRICKZ4BBhi86MA

### CODE REVIEW

**Review by ShyamsundarR — COMMENTED** [review]
@BenamarMk initial comments. Some are unrelated to the patch, so feel free to fix them up later or in a separate PR

File: api/v1alpha1/applicationvolumereplication_types.go [hunk 1]
```diff
@@ -20,32 +20,36 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// Action that will be taken by Ramen when a subscription is paused
+// DRAction that will be taken by Ramen when a subscription is paused
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply api/v1alpha1/applicationvolumereplication_types.go:23]: done

File: controllers/applicationvolumereplication_controller.go [hunk 2]
```diff
@@ -195,74 +190,237 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
...
+		return ctrl.Result{}, err
+	}
+
+	clonedPlRule, err := r.getOrClonePlacementRule(ctx, avr, userPlRule)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Make sure that we give time to the cloned PlacementRule to run and
+	// produces decisions
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/applicationvolumereplication_controller.go]: Status is not a pointer. and `len` should be fine with nil slice.  Just return 0

File: api/v1alpha1/applicationvolumereplication_types.go [hunk 3]
```diff
@@ -20,32 +20,36 @@ import (
...
 // +kubebuilder:validation:Enum=Failover;Failback
-type Action string
+type DRAction string
 
-// These are the valid values for Action
+// These are the valid values for DRAction
 const (
 	// Failover, restore PVs to the new home cluster, and unpause subscription
```
[the remark this answers is not in the document]
Comment (BenamarMk) [reply api/v1alpha1/applicationvolumereplication_types.go]: done

**Comment by BenamarMk** [comment]
I have updated this PR to simply allow others to get it and use it.  This may not end up being merged.  We are going to rework the Failover/Failback to remove the embiguity.  We will follow a simple approach.  Here what we are planning to do.
1. Rename AVR to a name that will be chosen from [this survey](https://docs.google.com/forms/d/e/1FAIpQLSeN9frnkDDy4bd9mI0R8jzYf-PVM9zpShVN2wuchZOZPz4FUg/viewform?vc=0&c=0&w=1&flr=0&usp=mail_form_link)
2. Add to AVR two more fields, PreferredCluster and TargetCluster
3. Add one more action to DRAction, Relocate
4. During Initial deployment: always use the PreferredCluster for the Home Cluster
3. When Action is set to Failover: always use the TargetCluster to failover to
4. When Action is set to Failback: always use the PreferredCluster to Failback
5. Action is set to Relocate: always use the TargetCluster to Relocate

The PreferredCluster is an immutable and mandatory field. We will error out if it does not exist. @ShyamsundarR, however, also indicated that we may need to allow modification to the PreferredCluster in the case of permanent relocation.

**Comment by BenamarMk** [comment]
No long valid PR.  Use PR #69

**Pull request closed without merging**